cmd/smoketests/main.go

```diff
@@ -0,0 +1,205 @@
+// Copyright (C) 2019 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// The smoketests command runs a series of smoke tests.
+package main
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/google/gapid/core/app"
+	"github.com/google/gapid/core/log"
+)
+
+var (
+	path = flag.String("traces", "traces", "The directory containing traces to run smoke tests on")
+)
+
+func main() {
+	app.ShortHelp = "smoketests runs a series of smoke tests"
+	app.Name = "smoketests"
+	app.Run(run)
+}
+
+func run(ctx context.Context) error {
+
+	// Check argument
+	filemode, err := os.Lstat(*path)
+	if err != nil {
+		return err
+	}
+	if !filemode.IsDir() {
+		return errors.New("Path given in argument is not a directory")
+	}
+	tracedir := *path
+
+	// Record starting working directory
+	startwd, err := os.Getwd()
+	if err != nil {
+		return err
+	}
+
+	// Create temporary directory
+	tmpdir, err := ioutil.TempDir(startwd, "smoketests.")
+	if err != nil {
+		return err
+	}
+
+	// For each trace, run gapit tests
+	if !filepath.IsAbs(tracedir) {
+		tracedir = filepath.Join(startwd, tracedir)
+	}
+	traces, err := ioutil.ReadDir(tracedir)
+	atLeastOneTraceFound := false
+	nbErr := 0
+
+	for _, t := range traces {
+
+		// Filter out non-trace files
+		if t.IsDir() {
+			continue
+		}
+		trace := t.Name()
+		if !strings.HasSuffix(trace, ".gfxtrace") {
+			continue
+		}
+
+		atLeastOneTraceFound = true
+
+		// Run smoke tests on trace under temporary directory
+		tracewd := filepath.Join(tmpdir, trace)
+		tracepath := filepath.Join(tracedir, trace)
+		if err := os.Mkdir(tracewd, 0777); err != nil {
+			return err
+		}
+		os.Chdir(tracewd)
+		err = testTrace(ctx, &nbErr, tracepath)
+		if err != nil {
+			return err
+		}
+		os.Chdir(startwd)
+	}
+
+	if !atLeastOneTraceFound {
+		return errors.New("No file ending with '.gfxtrace' found in trace directory")
+	}
+
+	if nbErr > 0 {
+		errStr := "error"
+		if nbErr > 1 {
+			errStr = "errors"
+		}
+		log.I(ctx, "%d %s found, see logs in %s", nbErr, errStr, tmpdir)
+	} else {
+		// Temporary directory is removed only when there are no errors
+		os.RemoveAll(tmpdir)
+	}
+	return nil
+}
+
+func testTrace(ctx context.Context, nbErr *int, tracepath string) error {
+	trace := filepath.Base(tracepath)
+	var err error
+	err = gapit(ctx, nbErr, "commands", tracepath)
+	if err != nil {
```

Comment:
nit: It is more idiomatic to keep the scopes of variables as short lived as possible.
This then becomes:

```
if err := gapit(ctx, nbErr, "commands", tracepath); err != nil {
```

Reply:
Thanks, this pushed me to refactor testTrace to make it easier to understand which commands are run.